epi/epi_functions.py

```diff
@@ -0,0 +1,285 @@
+
+    try:
+        xy = pd.read_excel(
+            xlsx,
+            header=0,
+            sheet_name="tb4",
+            skiprows=[0, 1],
+            parse_dates=True,
+            names=custom_column_names,
+            usecols=custom_column_indices,
+            converters=dict((i, int) for i in range(len(custom_column_indices)) if i > 0),
+            skipfooter=1,  # skips the SKUPAJ row at the end
+        )
+        xy["datum"] = pd.to_datetime(xy["datum"])
+    except XLRDError as xe:
+        raise XLRDError(
+            f"Unable to find sheet named tb4. Please examine {xlsx} if the sheet name has been spelled correctly."
+        )
+    except ValueError as ve:
+        raise ValueError("Chances are that one of your 'integers' is not an integer after all. Check data for "
+                         "incorrect entries.")
+
+    logger.debug(f"Successfully imported file {xlsx}.")
+
+    # 2: data integrity check
+    #   - expect first column to be date and monotinically increasing/decreasing
+    if not xy[xy["datum"].diff().dt.days < 0].empty:
+        raise ValueError(
+            "Datums are monotonically increasing. Please check the datum column to " "see if dates are in order."
+        )
+
+    #   - expect columns 2:11 to be counts for age groups 0-85+ for men
+    #   - expect columns 13:22 to be counts for age groups 0-85+ for women
+    # This is resolved on import using `converters`.
+
+    #   - all values for columns 2:22 should be >= 0 or NA (empty)
+    expect_counts = xy.iloc[:, 1 : len(xy.columns)]
+    expect_counts_gezero = expect_counts >= 0
+    expect_counts_null = expect_counts.isnull()
+    counts_or_nan = expect_counts_gezero.__or__(expect_counts_null)
+
+    bad_indices = xy.loc[~counts_or_nan.all(axis=1)]
+    if len(bad_indices) > 0:
+        raise ValueError(f"Some counts are not greater than 0 or are not NaN. Hint:\n\n" f"{bad_indices}")
+
+    logger.debug(f"Found no negative counts.")
+
+    xyall = pd.DataFrame({"datum": pd.date_range(start=xy["datum"].min(), end=xy["datum"].max())})
+
+    # 3: missing dates
+    #   - if any missing dates, ask user if assuming NA is ok
+    #     - if ok, add missing dates NA for all columns
+    #     - if not, abort
+    if xyall.shape[0] != xy.shape[0]:
+        logger.debug("Imputing missing data.")
+        xy = xyall.merge(xy, how="left", on="datum")
+
+    # Add 0 to missing fieds in order for the cumsum function to output expected result.
+    # If this is not done, the missing values create a gap in cumulated sum.
+    xy = xy.fillna(0)
+
+    # 4: calculate sums
+    #   - calculate cumulative sums per age group per gender (20 new variables)
+    cs = xy.iloc[:, 1 : len(xy.columns)].cumsum(axis=None)
+    #   - calculate row-wise age group sums (irrespective of gender) of cumulative
+    #   sums (10 new variables)
+    row_group_sums = []  # row-wise group sums
+    for age_group in age_groups:
+        tmp_cs = cs.filter(regex=age_group)
```

Comment:
I should probably run the code to test (I admit I didn't), but I'm curious what this does when `age_group` is equal to `85+`.

PS: even if it matches 855555, we likely don't care as we won't find this in the document => no need to change anything :)

Reply:
Column names are hard coded and I don't expect to have any other group starting with 85. Using the pattern, next one would be 95-105. :)